File: chronicle-pc/pipeline/chronicle_pipeline/note_frontmatter.py
"""Convention-complete frontmatter helpers for knowledge note creates.

Ensures ``created`` / ``updated`` / ``type`` / ``title`` per ``_system/conventions.md``
so empty creates are not bare markdown. Existing keys are preserved (except
``updated``, which is always stamped on create).
"""

from __future__ import annotations

import re
from datetime import date

_FRONTMATTER_RE = re.compile(r"\A---\r?\n(.*?)\r?\n---\r?\n?", re.DOTALL)

_DEFAULT_KEY_ORDER = ("title", "created", "updated", "type", "tags", "aliases")


def today_iso() -> str:
    return date.today().isoformat()


def parse_frontmatter(text: str) -> tuple[dict[str, str], str]:
    """Return (frontmatter dict, body after the closing fence)."""
    m = _FRONTMATTER_RE.match(text)
    if not m:
        return {}, text
    meta: dict[str, str] = {}
    for line in m.group(1).splitlines():
        if ":" not in line:
            continue
        key, _, val = line.partition(":")
        meta[key.strip()] = val.strip()
    return meta, text[m.end() :]


def format_frontmatter(fm: dict[str, str]) -> str:
    ordered: list[tuple[str, str]] = []
    seen: set[str] = set()
    for key in _DEFAULT_KEY_ORDER:
        if key in fm:
            ordered.append((key, fm[key]))
            seen.add(key)
    for key, val in fm.items():
        if key not in seen:
            ordered.append((key, val))
    lines = [f"{k}: {v}" for k, v in ordered]
    return "---\n" + "\n".join(lines) + "\n---"


def _set_key(fm: dict[str, str], key: str, value: str, *, overwrite: bool) -> None:
    lower_map = {k.lower(): k for k in fm}
    existing = lower_map.get(key.lower())
    if existing is not None:
        if not overwrite and fm[existing].strip() != "":
            return
        if existing != key:
            del fm[existing]
    fm[key] = value


def ensure_create_frontmatter(
    content: str,
    *,
    title: str | None = None,
    note_type: str = "note",
    today: str | None = None,
) -> str:
    """Fill missing create-time frontmatter; stamp ``updated`` to *today*."""
    day = today or today_iso()
    fm, body = parse_frontmatter(content)

    cleaned_title = (title or "").strip()
    if cleaned_title:
        _set_key(fm, "title", cleaned_title, overwrite=False)
    _set_key(fm, "created", day, overwrite=False)
    _set_key(fm, "updated", day, overwrite=True)
    _set_key(fm, "type", note_type, overwrite=False)
    _set_key(fm, "tags", "[]", overwrite=False)

    if body and not body.startswith("\n"):
        body = "\n" + body
    if not body:
        body = "\n"
    return f"{format_frontmatter(fm)}{body}"

File: chronicle-android/app/src/main/java/com/chronicle/app/NoteFrontmatter.kt
package com.chronicle.app

import java.time.LocalDate

/**
 * Minimal YAML frontmatter helpers for knowledge notes.
 * Auto-bumps `updated` on save; ensureCreateFrontmatter on create
 * (mirrors PC conventions: created/updated/type/title).
 */
object NoteFrontmatter {
    private val FRONTMATTER_BLOCK =
        Regex("""^---\r?\n([\s\S]*?)\r?\n---\r?\n?""")

    private val DEFAULT_KEY_ORDER =
        listOf("title", "created", "updated", "type", "tags", "aliases")

    /** ISO date for frontmatter fields. */
    fun todayDate(): String = LocalDate.now().toString()

    /** Body after YAML frontmatter (Timeline rollups); unchanged if no FM block. */
    fun stripFrontmatter(content: String): String {
        val match = FRONTMATTER_BLOCK.find(content) ?: return content
        return content.substring(match.range.last + 1).trimStart('\n', '\r')
    }

    /** Replace or insert `updated: YYYY-MM-DD` in an existing frontmatter block. */
    fun bumpUpdated(content: String, date: String = todayDate()): String {
        val match = FRONTMATTER_BLOCK.find(content) ?: return content
        val bodyStart = match.range.last + 1
        val fmText = match.groupValues[1]
        val updatedLine = Regex("""^updated:\s*.+$""", RegexOption.MULTILINE)
        val newFm = if (updatedLine.containsMatchIn(fmText)) {
            fmText.replace(updatedLine, "updated: $date")
        } else {
            fmText.trimEnd() + "\nupdated: $date"
        }
        return content.substring(0, match.range.first) +
            "---\n$newFm\n---\n" +
            content.substring(bodyStart)
    }

    /** Expand `{{title}}` / `{{date}}` placeholders in a template body. */
    fun applyTemplatePlaceholders(template: String, title: String, date: String = todayDate()): String =
        template
            .replace("{{title}}", title)
            .replace("{{date}}", date)

    /**
     * Fill missing create-time frontmatter per conventions.
     * Existing keys preserved; `updated` always stamped to [date].
     */
    fun ensureCreateFrontmatter(
        content: String,
        title: String? = null,
        type: String = "note",
        date: String = todayDate(),
    ): String {
        val match = FRONTMATTER_BLOCK.find(content)
        val fm = linkedMapOf<String, String>()
        val body: String
        if (match != null) {
            for (line in match.groupValues[1].lines()) {
                val idx = line.indexOf(':')
                if (idx <= 0) continue
                fm[line.substring(0, idx).trim()] = line.substring(idx + 1).trim()
            }
            body = content.substring(match.range.last + 1)
        } else {
            body = content
        }

        fun setKey(key: String, value: String, overwrite: Boolean) {
            val existing = fm.keys.firstOrNull { it.equals(key, ignoreCase = true) }
            if (existing != null) {
                if (!overwrite && fm[existing]!!.isNotBlank()) return
                if (existing != key) fm.remove(existing)
            }
            fm[key] = value
        }

        val cleaned = title?.trim().orEmpty()
        if (cleaned.isNotEmpty()) setKey("title", cleaned, overwrite = false)
        setKey("created", date, overwrite = false)
        setKey("updated", date, overwrite = true)
        setKey("type", type, overwrite = false)
        setKey("tags", "[]", overwrite = false)

        val ordered = linkedMapOf<String, String>()
        for (k in DEFAULT_KEY_ORDER) {
            fm[k]?.let { ordered[k] = it }
        }
        for ((k, v) in fm) {
            if (k !in ordered) ordered[k] = v
        }
        val fmBlock = ordered.entries.joinToString("\n") { "${it.key}: ${it.value}" }
        var nextBody = body
        if (nextBody.isNotEmpty() && !nextBody.startsWith("\n")) {
            nextBody = "\n$nextBody"
        }
        if (nextBody.isEmpty()) nextBody = "\n"
        return "---\n$fmBlock\n---$nextBody"
    }
}
